cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -221,14 +225,42 @@ protected void startUp() throws Exception {
         SparkRuntimeEnv.setProperty(key, sparkDefaultConf.getProperty(key));
       }
 
-      if (isLocal) {
+      if (masterEnv != null) {
+        // Only add cconf, hconf for master environment
+        localizeResources.add(new LocalizeResource(saveCConf(cConfCopy, tempDir)));
+        Configuration hConf = contextConfig.set(runtimeContext, pluginArchive).getConfiguration();
+        localizeResources.add(new LocalizeResource(saveHConf(hConf, tempDir)));
+
+        // Localize all the files from user resources
+        List<File> files = copyUserResources(context.getLocalizeResources(), tempDir);
+        for (File file : files) {
+          localizeResources.add(new LocalizeResource(file));
+        }
+        // Master environment would always contain this file.
+        // https://github.com/cdapio/cdap/blob/develop/cdap-spark-core3_2.12/src/k8s/Dockerfile#L46
+        jobFile = URI.create("local:/opt/cdap/cdap-spark-core/cdap-spark-core.jar");
+      } else if (isLocal) {
         // In local mode, always copy (or link if local) user requested resources
-        copyUserResources(context.getLocalizeResources(), tempDir);
+        List<File> files = copyUserResources(context.getLocalizeResources(), tempDir);
 
         File metricsConf = SparkMetricsSink.writeConfig(new File(tempDir, CDAP_METRICS_PROPERTIES));
         metricsConfPath = metricsConf.getAbsolutePath();
 
         extractPySparkLibrary(tempDir, extraPySparkFiles);
+
+        // Add localize resources for master environment
+        if (masterEnv != null) {
```

Comment:
seems like this will always be null, was the accidentally left in?

Reply:
Ah yes, unintentional.

Reply:
ah yeah, have to be careful to use the spark-dev profile